js/src/elements/star-rating.ts

```diff
@@ -0,0 +1,289 @@
+const starRatingTemplate = document.createElement('template');
+starRatingTemplate.id = 'star-rating-template';
+
+/* TODO: Consider replacing with constructable stylesheets when there is broader support for Firefox and Safari.
+https://web.dev/custom-elements-v1/
+https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet */
+
+starRatingTemplate.innerHTML = `
+<style>
+	*,
+	::before,
+	::after {
+		box-sizing: border-box;
+	}
+	
+	:host {
+		--star-color: var(--theme-primary-active);
+	
+		display: block;
+	}
+
+
+	:host([disabled]) {	
+		--star-color: var(--theme-text);
+
+		cursor: not-allowed;
+	}
+	
+	fieldset {
+		display: flex;
+        margin: 0;
+        border: none;
+        padding-block: 0.75rem;
+	}
+	
+	.star-container {
+		display: flex;
+	}
+
+	svg {
+		fill: none;
+		stroke: var(--star-color);
+	}
+
+	/* On hover, fill stars as default so that stars prior to hovered/checked star looks filled, then un-fill the following stars */
+
+	.star-container > input:not(.is-selected) ~ label svg,
+	.star-container > input + label:hover ~ label svg,
+	.star-container > input:checked ~ label svg {
+		fill: none;
+	}
+
+	.star-container:hover > input + label svg,
+	.star-container > input.is-selected + label svg,
+	.star-container input:focus-visible:not(:disabled) + label svg {
+		fill: var(--star-color);
+	}
+
+	fieldset:disabled .star-container:hover input:not(.is-selected) + label svg{ 
+		fill: none;
+	}
+	
+	#alert {
+		margin-inline-start: 0.5rem;
+		line-height: 1;
+	}
+	
+	.star-container > label {
+		padding-inline: 0.125em;
+	}
+	
+	[id^="label-"] {
+		display: none;
+	}
+	
+	input:focus-visible + label {
+		border-radius: 0.5em;
+		outline: 3px dashed;
+	}
+	
+	input[value="1"]:checked ~ #alert #label-1,
+	input[value="2"]:checked ~ #alert #label-2,
+	input[value="3"]:checked ~ #alert #label-3,
+	input[value="4"]:checked ~ #alert #label-4,
+	input[value="5"]:checked ~ #alert #label-5 {
+		display: inline;
+	}
+
+	input[value="1"]:focus-visible ~ #alert #label-1,
+	input[value="2"]:focus-visible ~ #alert #label-2,
+	input[value="3"]:focus-visible ~ #alert #label-3,
+	input[value="4"]:focus-visible ~ #alert #label-4,
+	input[value="5"]:focus-visible ~ #alert #label-5 {
+		display: inline;
+	}
+	
+	/* override checked styles with hover styles */
+	
+	input:hover ~ #alert [id^="label-"] {
+		display: none !important;
+	}
+	
+	input[value="1"]:hover ~ #alert #label-1,
+	input[value="2"]:hover ~ #alert #label-2,
+	input[value="3"]:hover ~ #alert #label-3,
+	input[value="4"]:hover ~ #alert #label-4,
+	input[value="5"]:hover ~ #alert #label-5 {
+		display: inline !important;
+	}
+	
+	fieldset:disabled span,
+	input:disabled + label {
+		cursor: not-allowed;
+		pointer-events: none;
+	}
+  </style>
+
+<fieldset>
+	<slot name="legend" part="legend">Enter rating</slot>
+
+	<div class="star-container">
+		<input part="visually-hidden" type="radio" value="1" id="radio-1" />
```

Comment:
I opened up the accessibility tree display to see what the inputs' accessible name was that screenreaders would announce, and the accessible names are coming up empty. See:
![Screenshot of accessibility tree display for an input in the star rating widget. Although multiple labels are associated with the input, the accessible name is coming up as an empty string.](https://user-images.githubusercontent.com/18060369/166494792-b74c2a1b-6e91-4d6e-97eb-7cbc6fb786a3.PNG)

Each individual star input has two labels associated with it -- the star and the "Poor"/"Fair"/etc. rating, but I think the star label is winning out and it doesn't have any accessible text inside of it. This means that as screenreader users navigate between the inputs, they might not be told any information like which star they're on or what that rating means.

Reply:
(For anyone unfamiliar with the accessibility tree: I'm looking under the Computed Properties pane on the right. The computed name is `""`, which it's getting from two `label` elements)

Reply:
Might be worth disregarding - this works as intended in VoiceOver, so maybe fieldset works differently than I expected 😅

Reply:
Resolving this particular thread as the radios' legend names are now coming up in the accessibility tree.